src/typecheck/eq.rs

```diff
@@ -43,24 +43,61 @@
 //! times. For anything more complex, we return false.
 
 use super::*;
-use crate::eval;
-use crate::term::UnaryOp;
+use crate::{eval, term::UnaryOp};
 
 /// The maximal number of variable links we want to unfold before abandoning the check. It should
 /// stay low, but has been fixed arbitrarily: feel fee to increase reasonably if it turns out
 /// legitimate type equalities between simple contracts are unduly rejected in practice.
 pub const MAX_GAS: u8 = 8;
 
+/// Abstract over the term environment, which is represented differently in the typechecker and
+/// during evaluation.
+///
+/// The evaluation environment holds [`crate::eval::lazy::Thunks`], which are `Rc<RefCell<_>>`
+/// under the hood, while the term environment used during typechecking is just maps identifiers to
+/// a pair `(RichTerm, Environment)`. To have an interface that works with both,
+/// `TermEnvironment::get` has to take a closure representing the continuation of the task to do
+/// with the result instead of merely returning it.
+pub trait TermEnvironment: Clone {
```

Comment:
Because it may be stored in `GenericTypeWrapper<E>`, which itself has to be cloneable.

Reply:
Hmm, thinking about this again, you're probably right. It's `GenericTypeWrapper` that should impose an additional clone constraints on its parameter, I guess. On the other hand, it's not that farfetched to require `Clone` on something that is supposed to be an environment. Still, I'll move it to GenericTypeWrapper.

Reply:
Actually, I take that back. If I just add the constraint on `GenericTypeWrapper<E : TermEnvironment + Clone>`, I have to add that constraint in every single place, because all functions of `typecheck::eq` use `E` inside `GenericTypeWrapper<E>` transitively. Which means that `TermEnvironment` alone is never used, and is useless without the clone bound. I think we should keep it like it is.

Here is the diff, for the record:

```diff
diff --git a/src/typecheck/eq.rs b/src/typecheck/eq.rs
index 3708137e..28982f64 100644
--- a/src/typecheck/eq.rs
+++ b/src/typecheck/eq.rs
@@ -58,7 +58,7 @@ pub const MAX_GAS: u8 = 8;
 /// a pair `(RichTerm, Environment)`. To have an interface that works with both,
 /// `TermEnvironment::get` has to take a closure representing the continuation of the task to do
 /// with the result instead of merely returning it.
-pub trait TermEnvironment: Clone {
+pub trait TermEnvironment {
     fn get_then<F, T>(&self, id: &Ident, f: F) -> T
     where
         F: FnOnce(Option<(&RichTerm, &Self)>) -> T;
@@ -150,7 +150,7 @@ impl State {
     }
 
     /// Create a fresh unique rigid type variable.
-    fn fresh_cst<E: TermEnvironment>(&mut self) -> GenericTypeWrapper<E> {
+    fn fresh_cst<E: TermEnvironment + Clone>(&mut self) -> GenericTypeWrapper<E> {
         let result = self.var_uid;
         self.var_uid += 1;
         GenericTypeWrapper::Constant(result)
@@ -174,7 +174,7 @@ impl State {
 ///
 /// - `env`: an environment mapping variables to their definition (the second placeholder in a
 ///   `let _ = _ in _`)
-pub fn contract_eq<E: TermEnvironment>(
+pub fn contract_eq<E: TermEnvironment + Clone>(
     var_uid: usize,
     t1: &RichTerm,
     env1: &E,
@@ -186,7 +186,7 @@ pub fn contract_eq<E: TermEnvironment>(
 
 /// Decide type equality on contracts in their respective environment and given the remaining gas
 /// `gas`.
-fn contract_eq_bounded<E: TermEnvironment>(
+fn contract_eq_bounded<E: TermEnvironment + Clone>(
     state: &mut State,
     t1: &RichTerm,
     env1: &E,
@@ -349,7 +349,7 @@ where
 ///
 /// Require the rows to be closed (i.e. the last element must be `RowEmpty`), otherwise `None` is
 /// returned. `None` is returned as well if a type encountered is not row, or if it is a enum row.
-fn rows_as_map<E: TermEnvironment>(
+fn rows_as_map<E: TermEnvironment + Clone>(
     ty: &GenericTypeWrapper<E>,
 ) -> Option<HashMap<Ident, &GenericTypeWrapper<E>>> {
     let mut map = HashMap::new();
@@ -370,7 +370,7 @@ fn rows_as_map<E: TermEnvironment>(
 /// Require the rows to be closed (i.e. the last element must be `RowEmpty`), otherwise `None` is
 /// returned. `None` is returned as well if a type encountered is not row type, or if it is a
 /// record row.
-fn rows_as_set<E: TermEnvironment>(ty: &GenericTypeWrapper<E>) -> Option<HashSet<Ident>> {
+fn rows_as_set<E: TermEnvironment + Clone>(ty: &GenericTypeWrapper<E>) -> Option<HashSet<Ident>> {
     let mut set = HashSet::new();
 
     ty.iter_as_rows().try_for_each(|item| match item {
@@ -394,7 +394,7 @@ fn rows_as_set<E: TermEnvironment>(ty: &GenericTypeWrapper<E>) -> Option<HashSet
 /// all the rigid type variables encountered have been introduced by `type_eq_bounded` itself. This
 /// is why we don't need unique identifiers that are distinct from the one used during
 /// typechecking, and we can just start from `0`.
-fn type_eq_bounded<E: TermEnvironment>(
+fn type_eq_bounded<E: TermEnvironment + Clone>(
     state: &mut State,
     ty1: &GenericTypeWrapper<E>,
     env1: &E,
@@ -423,7 +423,7 @@ fn type_eq_bounded<E: TermEnvironment>(
                 rows1.is_some() && rows2.is_some() && rows1 == rows2
             }
             (AbsType::StaticRecord(tyw1), AbsType::StaticRecord(tyw2)) => {
-                fn type_eq_bounded_wrapper<E: TermEnvironment>(
+                fn type_eq_bounded_wrapper<E: TermEnvironment + Clone>(
                     state: &mut State,
                     tyw1: &&GenericTypeWrapper<E>,
                     env1: &E,
diff --git a/src/typecheck/mod.rs b/src/typecheck/mod.rs
index e7fefaa8..01a8c063 100644
--- a/src/typecheck/mod.rs
+++ b/src/typecheck/mod.rs
@@ -1240,7 +1240,7 @@ impl From<AbsType<Box<TypeWrapper>>> for TypeWrapper {
     }
 }
 
-impl<'a, E: TermEnvironment> Iterator for RowIterator<'a, GenericTypeWrapper<E>> {
+impl<'a, E: TermEnvironment + Clone> Iterator for RowIterator<'a, GenericTypeWrapper<E>> {
     type Item = RowIteratorItem<'a, GenericTypeWrapper<E>>;
 
     fn next(&mut self) -> Option<Self::Item> {
```

Reply:
Okay, personally, I think I would have made a different trait `CloneableTermEnvironment` which would require `TermEnvironment + Clone`, but I guess it's a matter of preference, and not a real issue that requires a change :/